### ISSUE TITLE
Full AssemblyLoadContext unloading whilst in debug mode

### ISSUE THREAD

**Issue by mterhorst** [dotnet/runtime#47106]
We use a collectible `AssemblyLoadContext` for loading/unloading plugins.

We delete the plugin folder after the unload, by checking the `WeakReference.IsAlive` that references the `AssemblyLoadContext`.

However, when the Visual Studio debugger on Windows is attached, the deleting of the folder results in a `UnauthorizedAccessException` `Access to the path is denied`. 

Consider the following [Gist](https://gist.github.com/mterhorst/49fbe0504418cc9e4c2cf7f9d416f498).

A breakpoint on line 24 gives us no problems.

A breakpoint on line 28 (which is after the `AssemblyLoadContext.Unload()`) will throw that exception. The `WeakReference.IsAlive` reports `false`.

**Closing**

Can we conclude that the `AssemblyLoadContext` is indeed unloaded and that another process is keeping the dll file in use? If so, are there suggestions on how we can reliably remove plugins (including the removal of the dll) whilst in debug mode?

**Comment by vitek-karas** [comment]
You can check with for example [Process Explorer](https://docs.microsoft.com/en-us/sysinternals/downloads/process-explorer) it has "Find file handles" function which given the name of the file can tell you which process has it opened. I would guess it's the debugger... 
@dotnet/dotnet-diag - do we tell the debugger to close assemblies when they're unloaded?

**Comment by mterhorst** [comment]
I checked with Process Explorer and it shows that `devenv.exe` still uses the file after `AssemblyLoadContext.Unload()`.

**Comment by noahfalk** [comment]
> @dotnet/dotnet-diag - do we tell the debugger to close assemblies when they're unloaded?

We do deliver assembly unload and module unload notifications to VS but I doubt the debugger makes any promises that all file handles will be released as soon as that notification is delivered. "Debugger" here includes the library mscordbi.dll which is owned by our (.NET) team in addition to any code owned by VS debugger team proper. There is lots of complicated COM reference counting in that code and there have definitely been bugs found over the years that held reference counts higher than they would ideally be. Cases where we held the file locks after the debuggee process had exited get lots of attention because that stops people from rebuilding. However cases that hold a file lock after module unload while the debuggee process is still running is a much narrower case that only affects apps that dynamically unload plugins. It wouldn't surprise me if those issues either went unnoticed or unprioritized given how much less common it is. Feel free to re-assign it to me.

@mterhorst - Is it possible for you to work around this, for example by ignoring failures to delete the assemblies when operating under a debugger? I know its probably not what you were hoping to hear.

**Comment by janvorli** [comment]
I am aware of this issue. Many people have been hitting this in the past, some of them thinking unloading doesn't work until I've asked them to try without Visual Studio.

**Comment by noahfalk** [comment]
@janvorli - are you aware if anyone ever peeled the onion beyond knowing it was VS? Although I can't say I am surprised this was the first time I've seen a concrete example.

**Comment by janvorli** [comment]
I don't think it was ever investigated from the debugger side of things. The easiest way to repro it is to use my unloadability sample at https://github.com/dotnet/samples/tree/master/core/tutorials/Unloading and add `File.Delete(pluginFullPath);` at the end of the `Main` method in the `Program.cs`. It fails to delete the file under the debugger, but succeeds when running without it.

**Comment by mterhorst** [comment]
> @mterhorst - Is it possible for you to work around this, for example by ignoring failures to delete the assemblies when operating under a debugger? I know its probably not what you were hoping to hear.

It is certainly possible to workaround this, because this can only happen when the app is compiled in debug mode, which we only use while we are developing the application. So a few comments in the code will tell the other devs to be aware of the potential issue when using debugger breakpoints.

Somehow I think the [https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability) document should mention the relation between ALC unload and the VS debugger.

**Comment by janvorli** [comment]
@mterhorst the issue has nothing to do with whether the app is built in debug or release mode. Even if you run Release build under the Visual Studio debugger, you'd hit the same issue.

**Comment by noahfalk** [comment]
Thanks @mterhorst, glad to hear that. If needed code can check if it is running under the debugger: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.debugger.isattached?view=net-5.0
